src/Shared/Dictionary/SmallCapacityDictionary.cs

```diff
@@ -0,0 +1,760 @@
+            }
+
+            if (ContainsKeyCore(key))
+            {
+                return false;
+            }
+
+            EnsureCapacity(Count + 1);
+            _arrayStorage[Count] = new KeyValuePair<TKey, TValue>(key, value);
+            _count++;
+            return true;
+        }
+
+        /// <inheritdoc />
+        public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_backup != null)
+            {
+                return _backup.TryGetValue(key, out value);
+            }
+
+            return TryFindItem(key, out value);
+        }
+
+
+        [DoesNotReturn]
+        private static void ThrowArgumentNullExceptionForKey()
+        {
+            throw new ArgumentNullException("key");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnsureCapacity(int capacity)
+        {
+            EnsureCapacitySlow(capacity);
+        }
+
+        private void EnsureCapacitySlow(int capacity)
+        {
+            if (_arrayStorage.Length < capacity)
+            {
+                if (capacity > DefaultArrayThreshold)
+                {
+                    _backup = new Dictionary<TKey, TValue>(capacity);
+                    foreach (var item in _arrayStorage)
+                    {
+                        _backup[item.Key] = item.Value;
+                    }
+
+                    // Don't use _count or _arrayStorage anymore
+                    // TODO clear arrays here?
+                }
+                else
+                {
+                    capacity = _arrayStorage.Length == 0 ? DefaultArrayThreshold : _arrayStorage.Length * 2;
+                    var array = new KeyValuePair<TKey, TValue>[capacity];
+                    if (_count > 0)
+                    {
+                        Array.Copy(_arrayStorage, 0, array, 0, _count);
+                    }
+
+                    _arrayStorage = array;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int FindIndex(TKey key)
+        {
+            // Generally the bounds checking here will be elided by the JIT because this will be called
+            // on the same code path as EnsureCapacity.
+            var array = _arrayStorage;
+            var count = _count;
+
+            for (var i = 0; i < count; i++)
+            {
+                if (_comparer.Equals(array[i].Key, key))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryFindItem(TKey key, out TValue? value)
+        {
+            var array = _arrayStorage;
+            var count = _count;
+
+            // Elide bounds check for indexing.
```

Comment:
I believe that won't happen. `i` must be checked against array length in order to allow the JIT to elide the bound checks.

Same below.

Reply:
@JamesNK ?

Reply:
I'd trust @gfoidl over myself when it comes to JIT optimizations 😄

Reply:
I just removed the comments for now.

Reply:
Write these methods span-based as
```c#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private bool TryFindItem(TKey key, out TValue? value)
{
    var storage = _arrayStorage.AsSpan(0, _count);

    foreach (ref var item in storage)
    {
        if (_comparer.Equals(item.Key, key))
        {
            value = item.Value;
            return true;
        }
    }

    value = default;
    return false;
}
```
to elide the bound-checks. And it produces less code in total.

Reply:
I tried this and it was _significantly slower_:

Before:
```
| OneValue_SmallDict_Get |   108.3 ns |  29.01 ns |    77.94 ns |   100.0 ns | 9,230,769.2 |     - |     - |     - |      48 B |
```

After:
```
| OneValue_SmallDict_Get | 2,121.1 ns | 379.4 ns | 1,100.8 ns | 1,650.0 ns |   471,445.9 |     - |     - |     - |      48 B |
```

Reply:
Oh, I'll have a look what's going on here.